### PR TITLE
internal/store: delete resources

### PR DESCRIPTION
**Pull request by mhilton**
Add support in the store for deleting resources. This has the same
semantics as deleting entities.

### CODE REVIEW

File: internal/charmstore/resources.go [hunk 1]
```diff
@@ -137,6 +137,74 @@ func mapRevisions(resourceRevisions []mongodoc.ResourceRevision) map[string]int
 	return revisions
 }
 
+// DeleteResource deletes the resource with the given id from the store.
+// If the resource is the currently published revision for any channel or
+// the last revision for base entity and resource name an error will be
+// returned with an ErrForbidden cause.
```
Comment (cmars) [inline internal/charmstore/resources.go:143]: A 404 or 410 seems more appropriate, but as long as it's well defined and works well with the existing clients, juju, etc, it's a win.

File: internal/charmstore/resources.go [hunk 2]
```diff
@@ -137,6 +137,74 @@ func mapRevisions(resourceRevisions []mongodoc.ResourceRevision) map[string]int
...
+	if err != nil {
+		return errgo.Mask(err)
+	}
+
+	var res *mongodoc.Resource
+	for _, r := range resources {
+		if r.Revision == rev.Revision {
+			r := r
+			res = &r
+		}
```
Comment (cmars) [inline internal/charmstore/resources.go:161]: break?

File: internal/charmstore/resources.go [hunk 3]
```diff
@@ -137,6 +137,74 @@ func mapRevisions(resourceRevisions []mongodoc.ResourceRevision) map[string]int
...
+		{"name", rev.Name},
+	}).All(&resources)
+	if err != nil {
+		return errgo.Mask(err)
+	}
+
+	var res *mongodoc.Resource
+	for _, r := range resources {
+		if r.Revision == rev.Revision {
+			r := r
```
Comment (cmars) [inline internal/charmstore/resources.go]: Would using the slice index be a more concise and direct way to get this done? Something like:

```
for i := range resources { ...
  if resources[i].Revision == rev.Revision {
    res = &resources[i]
    break
...
```

File: internal/charmstore/resources.go [hunk 4]
```diff
@@ -137,6 +137,74 @@ func mapRevisions(resourceRevisions []mongodoc.ResourceRevision) map[string]int
...
+	// Remove the resource.
+	err = s.DB.Resources().Remove(bson.D{
+		{"baseurl", mongodoc.BaseURL(&id.URL)},
+		{"name", rev.Name},
+		{"revision", rev.Revision},
+	})
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			// Someone else got there first.
+			err = params.ErrNotFound
```
Comment (cmars) [inline internal/charmstore/resources.go:200]: This reassignment is weird to me, but maybe it's necessary or idiomatic in this project or an errgo thing?

**Review by cmars — APPROVED** [review]

**Review by frankban — APPROVED** [review]
Looks good with notes from Casey. Thanks!

**Pull request merged**